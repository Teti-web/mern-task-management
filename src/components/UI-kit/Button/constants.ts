export enum IconStyleButton {
  ICON_LEFT = "icon-left",
  ICON_RIGHT = "icon-right",
  ICON_ONLY = "icon-only",
}
