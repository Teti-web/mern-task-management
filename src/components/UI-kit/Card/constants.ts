export const CARD_TEST_ID = "cardComponent";
