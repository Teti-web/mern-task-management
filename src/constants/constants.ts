export enum Size {
  SMALL = "small",
  MEDIUM = "medium",
  LARGE = "large",
  CIRCLE = "circle",
}

export enum Style {
  PRIMARY = "primary",
  SECONDARY = "secondary",
  MINIMAL = "minimal",
  CIRCLE = "circle",
}
