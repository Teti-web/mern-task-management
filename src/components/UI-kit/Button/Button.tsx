import { Size, Style } from "@/constants/constants";
import { FC } from "react";
import { IconStyleButton } from "./constants";

export interface ButtonType {
  label: string;
  isLink: boolean;
  style: Style;
  size: Size;
  link?: string;
  isIcon?: boolean;
  iconStyle?: IconStyleButton;
  icon?: string;
  altIcon?: string;
  isDisabled?: boolean;
  onClick?: () => void;
}

const Button: FC<ButtonType> = ({
  label,
  isLink = true,
  link = "#",
  iconStyle,
  icon,
  altIcon = "Icon",
  onClick,
  isDisabled,
  style = Style.PRIMARY,
  size = Size.SMALL,
}) => {
  let className =
    "font-primary rounded-[10px] focus:border-[1px] focus:border-solid transition-all ease-in-out  duration-500";

  if (style === Style.PRIMARY) {
    className +=
      " text-white bg-primary-500 font-semibold hover:bg-primary-600 focus:border-primary-200 focus:bg-primary-500  disable:bg-primary-500  active:bg-primary-700 ";
  } else if (style === Style.MINIMAL) {
    className +=
      " text-secondinary-300 font-semibold bg-transparent hover:text-secondinary-500 hover:bg-white active:bg-secondinary-100 active:text-secondinary-500 focus:border-secondinary-300 focus:text-secondinary-500 focus:bg-white";
  } else if (style === Style.SECONDARY) {
    className +=
      " text-secondinary-400 font-semibold bg-white border-[1px] border-secondinary-200 hover:text-secondinary-500 hover:border-secondinary-400 active:bg-secondinary-200 active:text-secondinary-400 focus:border-secondinary-200 focus:text-secondinary-500 focus:bg-white";
  }

  if (size === Size.SMALL) {
    className += " text-[12px] pt-[5px] pb-[5px] pl-[16px] pr-[16px]";
  } else if (size === Size.MEDIUM) {
    className += " text-[12px] pt-[9px] pb-[9px] pl-[16px] pr-[16px]";
  } else if (size === Size.LARGE) {
    className += " text-[16px] pt-[13px] pb-[13px] pl-[41px] pr-[41px]";
  }

  if (isDisabled) {
    className += " opacity-40 cursor-not-allowed ";
  }

  if (iconStyle === IconStyleButton.ICON_LEFT) {
    className += " flex items-center w-max gap-[4px]";
    if (isLink) {
      return (
        <a href={link} className={className} aria-disabled={isDisabled}>
          <img src={icon} alt={altIcon} /> {label}
        </a>
      );
    } else {
      return (
        <button onClick={onClick} className={className} disabled={isDisabled}>
          <img src={icon} alt={altIcon} /> {label}
        </button>
      );
    }
  } else if (iconStyle === IconStyleButton.ICON_RIGHT) {
    className += " flex items-center w-max gap-[4px]";
    if (isLink) {
      return (
        <a href={link} className={className} aria-disabled={isDisabled}>
          {label}
          <img src={icon} alt={altIcon} />
        </a>
      );
    } else {
      return (
        <button onClick={onClick} className={className} disabled={isDisabled}>
          {label}
          <img src={icon} alt={altIcon} />
        </button>
      );
    }
  } else if (iconStyle === IconStyleButton.ICON_ONLY) {
    if (isLink) {
      return (
        <a href={link} className={className} aria-disabled={isDisabled}>
          <img src={icon} alt={altIcon} />
        </a>
      );
    } else {
      return (
        <button onClick={onClick} className={className} disabled={isDisabled}>
          <img src={icon} alt={altIcon} />
        </button>
      );
    }
  }

  return (
    <button onClick={onClick} className={className} disabled={isDisabled}>
      {label}
    </button>
  );
};

export default Button;
