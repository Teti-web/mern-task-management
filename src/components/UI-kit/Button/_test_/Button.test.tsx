import "@testing-library/jest-dom";
import { render, cleanup } from "@testing-library/react";
import Button from "../Button";
import { Style, Size } from "@/constants/constants";
import { IconStyleButton, BUTTON_TEST_ID } from "../constants";

describe("Button component", () => {
  afterEach(cleanup);

  it("renders button as a link with icon on the left", () => {
    const { getByTestId, getByAltText, asFragment } = render(
      <Button
        label="Test Button"
        isLink={true}
        style={Style.MINIMAL}
        size={Size.SMALL}
        iconStyle={IconStyleButton.ICON_LEFT}
        icon="/path/to/icon.png"
        altIcon="Left Icon"
      />
    );

    const buttonLink = getByTestId(BUTTON_TEST_ID);
    expect(buttonLink).toBeInTheDocument();

    const icon = getByAltText("Left Icon");
    expect(icon).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button with only icon", () => {
    const { getByTestId, getByAltText, asFragment } = render(
      <Button
        label=""
        isLink={false}
        style={Style.MINIMAL}
        size={Size.SMALL}
        iconStyle={IconStyleButton.ICON_ONLY}
        icon="/path/to/icon.png"
      />
    );

    const button = getByTestId(BUTTON_TEST_ID);
    expect(button).toBeInTheDocument();

    const icon = getByAltText("Icon");
    expect(icon).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button without icon", () => {
    const { getByTestId, queryByAltText, asFragment } = render(
      <Button
        label="Test Button"
        isLink={false}
        style={Style.MINIMAL}
        size={Size.SMALL}
      />
    );

    const button = getByTestId(BUTTON_TEST_ID);
    expect(button).toBeInTheDocument();

    const icon = queryByAltText("Icon");
    expect(icon).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders button with primary style and medium size", () => {
    const { getByTestId, asFragment } = render(
      <Button
        label="Test Button"
        isLink={false}
        style={Style.PRIMARY}
        size={Size.MEDIUM}
      />
    );

    const button = getByTestId(BUTTON_TEST_ID);
    expect(button).toHaveClass(
      "text-[12px] pt-[9px] pb-[9px] pl-[16px] pr-[16px]"
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders button with  style minimal and LARGE size", () => {
    const { getByTestId, asFragment } = render(
      <Button
        label="Test Button"
        isLink={false}
        style={Style.MINIMAL}
        size={Size.LARGE}
      />
    );

    const button = getByTestId(BUTTON_TEST_ID);
    expect(button).toHaveClass(
      "text-secondinary-300 text-[16px] pt-[13px] pb-[13px] pl-[41px] pr-[41px]"
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("render button when disabled", () => {
    const { getByTestId, asFragment } = render(
      <Button
        label="Test Button"
        isLink={false}
        style={Style.PRIMARY}
        size={Size.SMALL}
        isDisabled={true}
      />
    );

    const button = getByTestId(BUTTON_TEST_ID);
    expect(button).toBeDisabled();
    expect(button).toHaveClass(" opacity-40 cursor-not-allowed ");
    expect(asFragment()).toMatchSnapshot();
  });
});
