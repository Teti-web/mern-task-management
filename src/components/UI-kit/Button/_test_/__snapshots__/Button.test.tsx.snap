// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component render button when disabled 1`] = `
<DocumentFragment>
  <button
    class="font-primary rounded-[10px] focus:border-[1px] focus:border-solid transition-all ease-in-out  duration-500 text-white bg-primary-500 font-semibold hover:bg-primary-600 focus:border-primary-200 focus:bg-primary-500  disable:bg-primary-500  active:bg-primary-700  text-[12px] pt-[5px] pb-[5px] pl-[16px] pr-[16px] opacity-40 cursor-not-allowed "
    data-testid="buttonComponent"
    disabled=""
  >
    Test Button
  </button>
</DocumentFragment>
`;

exports[`Button component renders button as a link with icon on the left 1`] = `
<DocumentFragment>
  <a
    class="font-primary rounded-[10px] focus:border-[1px] focus:border-solid transition-all ease-in-out  duration-500 text-secondinary-300 font-semibold bg-transparent hover:text-secondinary-500 hover:bg-white active:bg-secondinary-100 active:text-secondinary-500 focus:border-secondinary-300 focus:text-secondinary-500 focus:bg-white text-[12px] pt-[5px] pb-[5px] pl-[16px] pr-[16px] flex items-center w-max gap-[4px]"
    data-testid="buttonComponent"
    href="#"
  >
    <img
      alt="Left Icon"
      src="/path/to/icon.png"
    />
     Test Button
  </a>
</DocumentFragment>
`;

exports[`Button component renders button with  style minimal and LARGE size 1`] = `
<DocumentFragment>
  <button
    class="font-primary rounded-[10px] focus:border-[1px] focus:border-solid transition-all ease-in-out  duration-500 text-secondinary-300 font-semibold bg-transparent hover:text-secondinary-500 hover:bg-white active:bg-secondinary-100 active:text-secondinary-500 focus:border-secondinary-300 focus:text-secondinary-500 focus:bg-white text-[16px] pt-[13px] pb-[13px] pl-[41px] pr-[41px]"
    data-testid="buttonComponent"
  >
    Test Button
  </button>
</DocumentFragment>
`;

exports[`Button component renders button with only icon 1`] = `
<DocumentFragment>
  <button
    class="font-primary rounded-[10px] focus:border-[1px] focus:border-solid transition-all ease-in-out  duration-500 text-secondinary-300 font-semibold bg-transparent hover:text-secondinary-500 hover:bg-white active:bg-secondinary-100 active:text-secondinary-500 focus:border-secondinary-300 focus:text-secondinary-500 focus:bg-white text-[12px] pt-[5px] pb-[5px] pl-[16px] pr-[16px]"
    data-testid="buttonComponent"
  >
    <img
      alt="Icon"
      src="/path/to/icon.png"
    />
  </button>
</DocumentFragment>
`;

exports[`Button component renders button with primary style and medium size 1`] = `
<DocumentFragment>
  <button
    class="font-primary rounded-[10px] focus:border-[1px] focus:border-solid transition-all ease-in-out  duration-500 text-white bg-primary-500 font-semibold hover:bg-primary-600 focus:border-primary-200 focus:bg-primary-500  disable:bg-primary-500  active:bg-primary-700  text-[12px] pt-[9px] pb-[9px] pl-[16px] pr-[16px]"
    data-testid="buttonComponent"
  >
    Test Button
  </button>
</DocumentFragment>
`;

exports[`Button component renders button without icon 1`] = `
<DocumentFragment>
  <button
    class="font-primary rounded-[10px] focus:border-[1px] focus:border-solid transition-all ease-in-out  duration-500 text-secondinary-300 font-semibold bg-transparent hover:text-secondinary-500 hover:bg-white active:bg-secondinary-100 active:text-secondinary-500 focus:border-secondinary-300 focus:text-secondinary-500 focus:bg-white text-[12px] pt-[5px] pb-[5px] pl-[16px] pr-[16px]"
    data-testid="buttonComponent"
  >
    Test Button
  </button>
</DocumentFragment>
`;
